app/services/discovery_engine/sync/locking.rb

```diff
@@ -0,0 +1,69 @@
+module DiscoveryEngine::Sync
+  module Locking
+    class FailedToAcquireLockError < StandardError; end
+
+    # Redis key prefix for locks and versions
+    LOCK_KEY_PREFIX = "search_api_v2:sync_lock".freeze
+    VERSION_KEY_PREFIX = "search_api_v2:latest_synced_version".freeze
+
+    # Time-to-live for document locks if not explicitly released
+    DOCUMENT_LOCK_TTL = 30.seconds.in_milliseconds
+
+    # Options for Redlock client around handling retries
+    RETRY_COUNT = 10
+    RETRY_DELAY = 5.seconds.in_milliseconds
+    RETRY_JITTER = 5.seconds.in_milliseconds
+
+    # Locks a document while a critical section block is executed to avoid multiple workers
+    # competing to update the same document.
+    def with_locked_document(content_id, payload_version:, &critical_section)
```

Comment:
So the reason `Locking` hasn't got any tests of its own is that it's really only a mixin for deduplication and tidiness purposes so we don't have the same code in both `Put` and `Delete` (rather than a class in its own right that provides specific functionality to potential consumers outside this module).

It's only used in those two classes within the same module, and its end-to-end behaviour is completely tested in the tests for those classes.

We've got a similar pattern in a few other places where we use modules/mixins/concerns tactically to tidy things up (where it makes sense and doesn't lead to an endless game of chase-down-the-call-stack). Sadly there's no (or rather no _good_) way in Ruby's visibility/package semantics to mark a module as being internal to another module (or unexported, private, etc) 😞 

I've added a module comment to the file (which was missing anyway because naughty GOV.UK Rubocop has the top-level comment check turned off!) that should make it clearer that this isn't an API that anyone can just use and expect to be a thing in its own right!